### PR TITLE
OSGi-ify 3rd party libraries

### PR DESCRIPTION
**Pull request by J-N-K**
This is a POC for OSGi-ifying 3rd-party libraries. This makes it possible to install them manually from the karaf console (since no wrapping is needed anymore) and also helps with "hidden" dependencies (i.e. the non-OSGi libraries needs something which cannot be seen from the karaf feature check because the requirements are not defined).

This could also be moved to another repo.

### CODE REVIEW

File: osgiify/pom.xml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            </execution>
+          </executions>
+        </plugin>
+        <plugin>
+          <groupId>com.googlecode.addjars-maven-plugin</groupId>
+          <artifactId>addjars-maven-plugin</artifactId>
+          <version>1.0.5</version>
+          <executions>
+            <execution>
+			  <id>add-libs</id>
```
Comment (hakan42) [inline osgiify/pom.xml]: Strange tabs here 😄

Comment (J-N-K) [reply osgiify/pom.xml]: There are a lot of wrong indentation in that file.

**Comment by davidgraeff** [comment]
Eclipse and Apache have a maven repository each where they store osgi-fied libraries. But I guess to stay independent it might make sense to have an own repo.

**Comment by kaikreuzer** [comment]
> This could also be moved to another repo.

It definitely should not be a source (git) repo. In other cases, we already have uploaded them to our own bintray repo (see e.g. https://bintray.com/openhab/mvn/quercus-osgi), which makes them automatically available to the Maven build.

**Comment by J-N-K** [comment]
I think we should have one repo for all these libs, otherwise we would end up with 50+ additional repos in OH.

I would suggest to use this approach (starting with released binaries) ,otherwise we would have to maintain several different build systems. We could reduce the number of artifacts by making "special" OH bundles (i.e. excluding content that we don't use and bundling packages from different input bundles).

**Comment by kaikreuzer** [comment]
> I would suggest to use this approach (starting with released binaries)

Could you briefly elaborate what this approach exactly is? It isn't yet clear to me.
In general, I would hope that re-packaging such jars as bundles is a rare exception. If possible, we should always use existing bundles from other repos (such as Apache & Eclipse as @davidgraeff mentioned).
So which ones actually require us to re-package them? calimero, ok. What else?

**Comment by J-N-K** [comment]
If you look at the features.xml: all libraries that use wrap. These are often quite special (like the Bosch Indego library). I looked for osgi-versions during migration but did not find them.

**Comment by kaikreuzer** [comment]
Hm, that's quite a lot.
I'm not sure it is worth the effort - what's the problem with the wrapping (which is imho quite neat)? Just that the transitive dependencies cannot be seen by the Karaf feature verification?

**Comment by J-N-K** [comment]
1. Some libraries are not available from any repo at all (they are still in this repo). We could make this repo free of all binary content.

2. Non-OSGi bundles cannot be installed manually because wrap is not available on the karaf console.

**Comment by kaikreuzer** [comment]
1. Ideally, those libs should be provided by the repo where their source code resides. Moving them to our Bintray repo should only be a last resort (and we must make sure that it is clear, how it was built and how people can rebuild it for patches etc.).
2. But they can be installed as part of a feature, can't them? As for openHAB they are always only a dependency of a feature, I don't see any problem here.

**Comment by davidgraeff** [comment]
Kai you mention that you don't want many bundles in core because that causes increased loading time on embedded systems. You realize that "wrap" is performed during loading and that the source jar has to be unpacked, be edited and repacked again by wrap?

I don't see how that fits together. If I install one of the bindings with a dozen of wrapped dependencies that will take a lot of additional memory and time (and karaf has no progress report).

I only see this as an intermediate step to get the repo tycho free faster.

Regarding the second point: Is dropping a jar in the addons directory not a feature that we want to keep? Because most bindings have dependencies. Those need to be dropped into that directory as well (no problem if you tell people to do that), but those dependencies only work if they are OSGi bundles afaik.

If you can point people to bintray for downloading our wrapped bundles, that's at least a solution in contrast to saying: Sorry but that addon will not work in the addon directory.

**Comment by kaikreuzer** [comment]
>  If I install one of the bindings with a dozen of wrapped dependencies 

I agree, this is far from ideal. My assumption was that wrapping is rather the exception and only happens for certain users with specific bindings. But I'm absolutely ok if you want to get rid off it. My concern was rather that it becomes more complicated for people to use non-osgi dependencies as they have to do the additional step to manually wrap it and upload it to our bintray repo.

>  Is dropping a jar in the addons directory not a feature that we want to keep?

I consider the addons folder to be a remainder from openHAB 1 and not a feature for end users anymore, but rather for developers. The main reason to go for Karaf in 2.x was to replace the "user has to figure out what dependencies are needed" by Karaf features, which cleanly provide everything required.

To avoid us having to manually wrap and upload dependencies: Could we possibly set up a Maven pom, which depends on the non-osgi libraries, wraps them and publishes the osgi'ified versions to Bintray by a build job, similar to https://bintray.com/openhab/p2/openhab-deps-repo?

**Comment by J-N-K** [comment]
@kaikreuzer Actually the idea here was to simplify that process by providing infratsructure for the wrapping. Everything that would be necessary is dropping the jar in the project directory and amending the [pom.xml](https://github.com/openhab/openhab2-addons/pull/5503/files#diff-22de599a551dd26b72d002ccdb1ae932). `osgiify.import` is not even needed if everything shall be imported (which probably is the case for most libraries). I'm investigating if I can get bnd to automatically detect the `osgiify.library` property, that would be even simpler.

All other files in this PR are infrastructure and only needed once or automatically created.

We could provide a script that asks for the needed properties and the creates the project. I don't think this creates a high hurdle.

**Comment by kaikreuzer** [comment]
My point is that here in openhab2-addons, nothing should be required to do - a binding would simply have a dependency to the osgi-ified artifact that can be retrieved from Bintray.

All infrastructure to get it to Bintray should imho not be here in this repo, but rather in https://bintray.com/openhab/p2/openhab-deps-repo (or a similar repo).
The non-osgi versions must not be committed to the repo, but only be drawn in during the build (and for the rare cases, where we don't even have a non-osgi version available in Maven Central, we would manually upload it to openHAB Bintray). 

Does this make sense?

**Comment by J-N-K** [comment]
As said before: I think this repo should be source-only, so I fully agree that we should not do the wrapping here. I just selected this repo to provide a POC.

I‘ll try to figure out if I can get the wrapping working with a remote-dependency, but I failed when I tried last time.

**Comment by J-N-K** [comment]
This new version removes the binary and downloads it during build. The downloaded artifact is ignored by git. 

I see one problem for new contributions if we move this to another repo: the developer would need to raise PR to to repos: this one and the osgiify-one. The osgiify-PR needs to be merged before the other one even builds.

**Comment by davidgraeff** [comment]
@J-N-K That's not so bad actually. We don't want people encouraged to use 3rd party non-osgi libraries all the time.

Sometimes they just need one class out of a huge lib and if we have this two-step-involves-effort procedure hopefully people will think twice before introducing 3rd party non-osgi capable libraries. They can also always raise on issue on the 3rd party lib and ask for a manifest.mf file to be added. There are even gradle and maven plugins as we know.

**Comment by gsieben** [comment]
Guys, I just wanted to mention something at this point. I read in the "old" documentation about the libraries and that one should check if the libraries are already in use. But I didn't find where to check it. For instance I use the libraries of Pi4J. I don't think I'm the only one he needs for his binding. Maybe that should be considered in the new documentation. Best regards Gabriel

**Comment by davidgraeff** [comment]
I have created an issue: https://github.com/openhab/openhab-docs/issues/974

**Comment by J-N-K** [comment]
How do we move on here?

**Comment by davidgraeff** [comment]
We either wait for The Github Package Registry https://github.com/features/package-registry or use bintray. 

But in any way we will use https://github.com/openhab/openhab-deps-repo + travis for building the OSGi'fied libraries, I guess.

**Comment by J-N-K** [comment]
I have setup an extra-repository https://github.com/J-N-K/openhab-osgiify for this which can be integrated in the openhab organization later. 

This includes the two examples from this PR. By going this way we can get this repo free of any binary dependency and also remove the wrapping which is causing diverse problems.

WDYT?

**Comment by davidgraeff** [comment]
You should rather generate a github page as maven repo structure. Using git raw urls for binary data is an anti pattern. But yes I'm all for getting rid of any jars in here.

**Comment by wborn** [comment]
Maybe you can also have a look at how they do this with [Eclipse Orbit](https://git.eclipse.org/c/orbit/orbit-recipes.git/tree/)?

The biggest differences I see are that they just use osgi.bnd files which are more standard and properly [documented](https://bnd.bndtools.org/chapters/390-wrapping.html) instead of adding all this info to Maven POMs. It would also allow for more easily replacing Maven with something else one day. They also support wrapping multiple versions of an artifact.

**Comment by J-N-K** [comment]
I thought about something along these lines before (a .bnd file per bundle). I tried to simplify that, but of course it would be more flexible that way. 

This still does not solve the problem with the embedded libraries, we still can‘t download them from a repo.

**Comment by J-N-K** [comment]
I have created two proposals in the repo above. 

1. all instructions in the pom
2. only one instruction in the pom (the artifact to be downloaded) and a bnd-file for the rest. 

Thinking about it, I would go with (2), that's more flexible in case any special instructions need to be added.

**Comment by kaikreuzer** [comment]
> I have not included release management parts in the parent pom, I'm not an expert on that. Maybe someone with more experience in this field can take care of that.

I think we should publish it to our Artifactory (afaik, all the different versions could be a bit problematic for Bintray) - I can take care of adding the required infos once the rest is ready, assuming that the pom is created in a way that a "mvn deploy" deploys every single artifact.

**Comment by wborn** [comment]
What would be the versioning approach for the bundles? If bugfixes are done on these bundles they may need to be rereleased. It's a bad practice and results in issues if such bundles are re-released using the same version number. One approach could be to use a different qualifier for each artifact. Another could be to use another repo for each release. The last option is more similar to how Eclipse Orbit bundles are released to p2 repos. It would also allow for using the exact same version of the original artifacts.

**Comment by J-N-K** [comment]
I would keep the original versioning. It's confusing if we change that. The "bnd"-branch works in deploying to a local repo on my harddisk.

**Comment by kaikreuzer** [comment]
> I would keep the original versioning.

That was my assumption as well.

**Comment by wborn** [comment]
The [bnd](https://github.com/J-N-K/openhab-osgiify/tree/bnd) branch looks like a good start to me! I'd also like the artifacts to be build and stored using the existing openHAB infra (Jenkins/Artifactory). Using the original artifact version where possible would also be my first preference. :-) I also tried to keep all the bndtools info in bnd files when I added the infra for adding JAR files to add-on bundles (https://github.com/openhab/openhab2-addons/pull/5190). But I couldn't get it to work because IIRC some of the info came from the Maven files.

**Comment by J-N-K** [comment]
The Tchaikovsky library is the worst, I don’t think we have a similar one. It has dependencies we provide embedded in the library and needs additional native-code. That‘s why I picked it.

I would have preferred (e.g. in the calimero-case) to add dependencies to the pom and use the `unpack-dependencies` goal. That works fine, transitive dependencies can be excluded and it‘s the same we have now (except that adding dependencies is more „standard“ than the properties). However the maven-bnd-plugin then adds all classes for these transitive dependencies and we get a big jar with all dependencies of our original library embedded. Unfortunately it seems to be impossible to stop bnd from doing that. We can of course exclude them in the bnd-file but that is more error-prone.

**Comment by J-N-K** [comment]
@openhab/2-x-add-ons-maintainers Since we have real-world-problems with bindings not working at all (allplay, amazondashbutton, xmppclient) which could be easily solved, can we please move on here? We can provide scripts for the creation of new bundles later.

**Comment by J-N-K** [comment]
@davidgraeff that’s not what I meant. We should decide if we want to do it the way I proposed in the linked repo and then integrate that into openhab. No need to do it in this repo.

**Comment by davidgraeff** [comment]
> I have created two proposals in the repo above.
> 
> 2) .. only one instruction in the pom

Option 2

**Comment by wborn** [comment]
Option 2 still looks fine to me! I find the bnd files easier to read than all the properties in the POMs (which can result in long lines). It's also easier to figure out how the content will be used by looking at the bnd docs. I also find it sometimes quite complicated to turn something into an OSGi bundle. Maybe that gets easier with practice and newer bnd features.

Is the plan to first setup the whole infrastructure for these couple of bundles, transfer the repo to openHAB and then add more bundles?

**Comment by J-N-K** [comment]
I‘ll add all wrapped bundles and then we should transfer the repo. I‘ll keep the allplay out, so we don’t pollute this with binaries. We should discuss where we want to upload the libraries that are not available in any repo. 

I would suggest to upload them somewhere the way they are now and then only add the wrapping to our new repo.

**Comment by maggu2810** [comment]
Have you already considered adding the sources JAR file to the repository, too?
IMHO it should be present, otherwise debugging is much harder as the sources for this GAV cannot be found.

I used something like that (I assume you just need to change some variable names e.g. wrapped to origin etc.): https://github.com/maggu2810/osgi-ify.org.checkerframework.checker-qual/blob/fb2615b/pom.xml#L58-L110

**Comment by J-N-K** [comment]
@maggu2810, it would be easy if I would compile those bundles. But I‘m only wrapping them. I think most bundles have no sources bundle available.

**Comment by maggu2810** [comment]
At least for ever non OSGi artifact that is wrapped and taken from Maven Central there should also be a sources classified JAR.
I didn't have a look at your whole tree, just realized that you map mina related stuff and I am pretty sure that there are existing sources.

**Comment by J-N-K** [comment]
It works for all but one of the libraries I have now. I think this number might increase when we start with the currently embedded libraries (only few of them have sources). I have to disable the sources-goals in that case. I see two options:

1. Like I did here: https://github.com/J-N-K/openhab-osgiify/pull/2/files#diff-7edbb892df09184ff8d028ca732b8935
2. Add a zero-length file (`noSources`) to the directory and disable with a build profile

WDYT?

**Comment by maggu2810** [comment]
I assume you could use a profile based approach and activate a the profile if there is no file names (e.g.) `no-official-sources.profile-marker`

**Comment by J-N-K** [comment]
@maggu2810 this is what I suggested with Option 2, it‘s probably the easiest way.

**Comment by maggu2810** [comment]
> @maggu2810 this is what I suggested with Option 2, it‘s probably the easiest way.

Sorry, I missed to mention that I just rephrase your option 2 in my own words. It has been my intention but it seems it gets lost between my cafe break.

**Comment by J-N-K** [comment]
What is the best practice: onl update the feature with the osgi-ified bundles or update the pom-dependency, too?

**Comment by wborn** [comment]
I think using the bundles in both features and POMs makes the most sense. The same is done with the similar servicemix and orbit dependencies in core. That keeps it simple and you could even change the scope in the POM to compile so bnd uses the bundle dependencies in the Demo app.

File: pom.xml [hunk 2]
```diff
@@ -58,6 +58,13 @@
     </snapshotRepository>
   </distributionManagement>
 
+  <repositories>
+    <repository>
+      <id>osgiify</id>
+      <url>file://c:/testrepo</url>
```
Comment (martinvw) [inline pom.xml]: Just FTR this looks to be optimized for Windows ;-)

Comment (J-N-K) [reply pom.xml]: This needs to be replaced with the repo later, It's just my test-repo with the osgiified bundles. Commited accidentally :D

**Comment by J-N-K** [comment]
I‘ve completed both PR. I would suggest someone reviews the PR in the osgiify-repo, we transfer that to openhab-organization and then adjust this one to the new repo and merge it.

File: bundles/org.openhab.binding.minecraft/src/main/feature/feature.xml [hunk 3]
```diff
@@ -7,7 +7,7 @@
         <feature>openhab-transport-mdns</feature>
         <feature prerequisite="true">wrap</feature>
```
Comment (wborn) [inline bundles/org.openhab.binding.minecraft/src/main/feature/feature.xml]:
```suggestion
```

**Comment by J-N-K** [comment]
> 
> Is there a reason why some dependencies are still wrapped?
> 

- xmppclient:  already "unwrapped" by #5707, needs to be rebased to show up in this branch
- dlinksmarthome: under investigation for JAX-WS in my core-PR, I chose to address that later
- bluetooth: see #5706, there is still some problem with the native code, this should not block here
- javasound: will check 
- minecraft: forgot to remove the prerequisite

**Comment by J-N-K** [comment]
@kaikreuzer, how can we transfer the repo from J-N-K to openhab? I tried `transfer` in the settings section, but it says `you are not allowed to create repositories in the openhab organization`.

**Comment by kaikreuzer** [comment]
> how can we transfer the repo from J-N-K to openhab?

@J-N-K I've enabled a flag now that members can create repos - so I would hope that the transfer is possible now. Could you have a try?

**Comment by J-N-K** [comment]
Looks like it worked. Can you add the missing parts for the repo? `mvn deploy` works for my local repo, so probably not so hard to do.

**Comment by kaikreuzer** [comment]
Working on it...

**Comment by kaikreuzer** [comment]
Done with https://github.com/openhab/openhab-osgiify/commit/3b5ee6ccc9a31f135e2e5309220b30079544135b.

Everything successfully deployed to [Bintray](https://bintray.com/openhab/mvn/openhab-osgiify#files/org/openhab/osgiify).

We will also have to get rid off https://dl.bintray.com/j-n-k/openhab-3rdparty sooner or later, right? What would be your plan for that?

**Comment by J-N-K** [comment]
I don't have any plans there. We still have (I didn't address that until now) a lot of embedded libraries that are not uploaded anywhere. Can we transfer bintray-repos? If that is possible I would suggest that I add all  that we need and we transfer the repo later.

**Comment by martinvw** [comment]
> @J-N-K J-N-K removed the work in progress  label 2 days ago

You mean that is now ready to be reviewed and merged? Would be great to get this off your plate.

**Comment by J-N-K** [comment]
@martinvw exactly. The repo has been transferred and I adjusted this PR to pick it up.

**Comment by davidgraeff** [comment]
@J-N-K Could you write a sentence or two for the documentation? Can happen here in the comments and I'll apply it. But right now nobody except participants of this issue know about the repo.

**Comment by J-N-K** [comment]
What exactly should be part of the documentation? I can write a short „how to“, but that would be placed in the other repo.

**Comment by davidgraeff** [comment]
In the documentation belongs the note, that non-osgi dependencies need to be wrapped and uploaded / PR created on that other repo. Just that you know the full story when reading the documentation pages and get started.

**Pull request merged**